### PR TITLE
Brute force ERGM

### PR DESCRIPTION
**Pull request by orenri**
Implemented an ERGM that iterates the entire space of networks, and calculates stuff exactly, to be used as a benchmark. This is obviously tractable only for small networks.

### CODE REVIEW

File: ergm.py [hunk 1]
```diff
@@ -187,6 +201,127 @@ def sample_network(self, sampling_method="NaiveMetropolisHastings", seed_network
...
+    A class that implements ERGM by iterating over the entire space of networks and calculating stuff exactly (rather
+    than using statistical methods for approximating and sampling).
+    This is mainly for tests.
+    """
+    # The maximum number of nodes that is allowed for carrying brute force calculations (i.e. iterating the whole space
+    # of networks and calculating stuff exactly). This becomes not tractable above this limit.
+    MAX_NODES_BRUTE_FORCE = 6
+
+    @staticmethod
+    def construct_adj_mat_from_int(int_code: int, num_nodes: int, is_directed: bool) -> np.ndarray:
```
Comment (tomtalp) [inline ergm.py:216]: I would move this and `construct_int_from_adj_mat` to the utils.py file - seems like something that would be useful in other places as well

File: ergm.py [hunk 2]
```diff
@@ -187,6 +201,127 @@ def sample_network(self, sampling_method="NaiveMetropolisHastings", seed_network
...
+        super().__init__(n_nodes,
+                         network_statistics,
+                         is_directed,
+                         initial_thetas,
+                         initial_normalization_factor,
+                         seed_MCMC_proba)
+        self._all_weights = self._calc_all_weights()
+        self._normalization_factor = self._all_weights.sum()
+
+    # TODO: think of an alternative name? confusing with the static method
```
Comment (tomtalp) [inline ergm.py:259]: Why do we even need this function? instead of just calling `construct_adj_mat_from_int` directly?

File: ergm.py [hunk 3]
```diff
@@ -187,6 +201,127 @@ def sample_network(self, sampling_method="NaiveMetropolisHastings", seed_network
...
+        for i in range(space_size):
+            cur_adj_mat = self._get_adj_mat_from_int(i)
+            all_weights[i] = super().calculate_weight(cur_adj_mat)
+        return all_weights
+
+    def calculate_weight(self, W: np.ndarray):
+        adj_mat_idx = BruteForceERGM.construct_int_from_adj_mat(W, self._is_directed)
+        return self._all_weights[adj_mat_idx]
+
+    def sample_network(self, sampling_method="Exact", seed_network=None, steps=0):
```
Comment (tomtalp) [inline ergm.py:282]: why do you need a sample_network function in the BruteforceERGM class? doesn't seem to be used anywhere

File: ergm.py [hunk 4]
```diff
@@ -187,6 +201,127 @@ def sample_network(self, sampling_method="NaiveMetropolisHastings", seed_network
...
+        if sampling_method != "Exact":
+            raise ValueError("BruteForceERGM supports only exact sampling (this is its whole purpose)")
+
+        all_nets_probs = self._all_weights / self._normalization_factor
+        sampled_idx = np.random.choice(all_nets_probs.size, p=all_nets_probs)
+        return self._get_adj_mat_from_int(sampled_idx)
+
+    # TODO: these inherited ghost parameters are ugly (also in the sample_network method)
```
Comment (tomtalp) [inline ergm.py:290]: Move to __init__ ?

File: ergm.py [hunk 5]
```diff
@@ -187,6 +201,127 @@ def sample_network(self, sampling_method="NaiveMetropolisHastings", seed_network
...
+        def nll(thetas):
+            model = BruteForceERGM(self._n_nodes, self._network_statistics, initial_thetas=thetas,
+                                   is_directed=self._is_directed)
+            return np.log(model._normalization_factor) - np.log(model.calculate_weight(observed_network))
+
+        def nll_grad(thetas):
+            model = BruteForceERGM(self._n_nodes, self._network_statistics, initial_thetas=thetas,
+                                   is_directed=self._is_directed)
+            observed_features = model._network_statistics.calculate_statistics(observed_network)
+            all_probs = model._all_weights / model._normalization_factor
```
Comment (tomtalp) [inline ergm.py:301]: Why divide by `model._normalization_factor`?
Why not just do
` num_nets = len(model._all_weights)` or .size or something?

File: ergm.py (hunk 2 above)
Comment (tomtalp) [reply ergm.py:259]: (which you do with `construct_int_from_adj_mat` in `calculate_weight`)

**Comment by tomtalp** [comment]
`.idea` to .gitignore?

**Comment by orenri** [comment]
> `.idea` to .gitignore?

I think it's supposed to be a bad practice to put the whole directory in .gitignore, but I don't know why.
We can just put it there and handle future issues if there are.

File: ergm.py (hunk 3 above)
Comment (orenri) [reply ergm.py:282]: I didn't pay attention to the fact it is not in use...
Why shouldn't we implement it? Its logic is different than that of the parent ERGM class. If we have a class of exact calculations for small nets, doesn't it make sense to support the API of the parent class in full?

**Pull request merged**